Rebase for botorch/sampling/pathwise - Dr. Ruda Zhang, Taiwo Adebiyi, Sahran Ashoor  - University of Houson

seashoo opened the pull request:
<!--
Thank you for sending the PR! We appreciate you spending the time to make BoTorch better.

Help us understand your motivation by explaining why you decided to make this change.

You can learn more about contributing to BoTorch here: https://github.com/pytorch/botorch/blob/main/CONTRIBUTING.md
-->

## Motivation

Hi! I'm [Sahran Ashoor](https://www.linkedin.com/in/ashoorsahran/), an undergraduate research assistant working for the [Uncertainty Quantification Lab](https://uq.uh.edu/) at the University of Houston. I work under [Dr. Ruda Zhang](https://scholar.google.com/citations?user=ttmax_wAAAAJ&hl=en) and [Taiwo Adebiyi](https://scholar.google.com/citations?user=qJdhj-IAAAAJ&hl=en), both of whom having already spoken with Max Balandat regarding incorporating a rebase of botorch/sampling/pathwise (Largely written by [James. T. Wilson](https://scholar.google.com/citations?user=x2JSAQwAAAAJ&hl=en)). The changes included in this pull request are my best attempt at faithfully completing the change logs I was provided ([product_kernel_diff.txt](https://github.com/user-attachments/files/20035605/product_kernel_diff.txt)). 

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/main/CONTRIBUTING.md#pull-requests)?

Yes!

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

The entirety of the testing suite was ran through pytest. Through additional verification we've found that the logic may be offset, but we're hoping to work with you all and further validate these changes under the discretion of Dr. Zhang. Expect further communications directly from my lab that will provide more insight into the rebase. 

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/pytorch/botorch, and link to your PR here.)

N/A

Balandat commented:
Thanks @seashoo for the PR - this is a big one! It'll take me a bit of time to review this in detail, I plan to do a first higher-level pass this week. 

> Through additional verification we've found that the logic may be offset

What exactly does this mean?

TaiwoAdebiyi23 commented:
> Thanks @seashoo for the PR - this is a big one! It'll take me a bit of time to review this in detail, I plan to do a first higher-level pass this week.
> 
> > Through additional verification we've found that the logic may be offset
> 
> What exactly does this mean?

Hi @Balandat,

Thanks for the response! We've included a more detailed Project Overview section in the pull request description to clarify our validation approach. Specifically, we utilized the existing unit test files, which cover prior, updates, and posterior sampling, and ensured that all tests passed as part of this rebase. While these tests are comprehensive, we welcome any additional guidance you might have on further validating the code's robustness.

Balandat commented on botorch/sampling/pathwise/prior_samplers.py:
```diff
@@ -98,35 +98,66 @@ def _draw_kernel_feature_paths_ExactGP(
+@DrawKernelFeaturePaths.register(models.ModelListGP)
+def _draw_kernel_feature_paths_ModelListGP(
+    model: models.ModelListGP,
+    reducer: Optional[Callable[[List[Tensor]], Tensor]] = None,
```
```suggestion
    reducer: Callable[[List[Tensor]], Tensor] | None = None,
```

Balandat commented on botorch/sampling/pathwise/prior_samplers.py:
```diff
@@ -98,35 +98,66 @@ def _draw_kernel_feature_paths_ExactGP(
+        else model._task_feature
+    )
+
+    base_kernel = deepcopy(model.covar_module)
```
outside the scope of this PR, but let's add this here
```suggestion
    NOTE: May want to use a `ProductKernel` instead in `MultiTaskGP` 
    base_kernel = deepcopy(model.covar_module)
```

Balandat commented on botorch/sampling/pathwise/update_strategies.py:
```diff
@@ -48,22 +49,17 @@ def gaussian_update(
                                                 V
                                     "Gaussian pathwise update"
 
-    where `=` denotes equality in distribution, :math:`f \sim GP(0, k)`,
-    :math:`y \sim N(f(X), \Sigma)`, and :math:`\epsilon \sim N(0, \Sigma)`.
-    For more information, see [wilson2020sampling]_ and [wilson2021pathwise]_.
```
let's keep this

Balandat commented on botorch/sampling/pathwise/update_strategies.py:
```diff
@@ -84,16 +80,22 @@ def _gaussian_update_exact(
-        noise_values = torch.randn_like(sample_values).unsqueeze(-1)
-        noise_values = noise_covariance.cholesky() @ noise_values
-        sample_values = sample_values + noise_values.squeeze(-1)
+        # Generate noise values with correct shape
```
thanks for fixing this

Balandat commented on botorch/sampling/pathwise/update_strategies.py:
```diff
@@ -137,6 +139,117 @@ def _gaussian_update_ExactGP(
+
+    # Prepare product kernel
+    num_inputs = points.shape[-1]
+    task_index = (
+        num_inputs + model._task_feature
+        if model._task_feature < 0
+        else model._task_feature
+    )
```
It would be better to do this so we can always assume that it's positive and we don't have to do this custom handling. But I think this is ok for now as is, that change is beyond the scope of this PR. 
```suggestion
    # TODO: Changed `MultiTaskGP` to normalize the task feature in its constructor.
    task_index = (
        num_inputs + model._task_feature
        if model._task_feature < 0
        else model._task_feature
    )
```

hvarfner replied:
Should be resolved in https://github.com/pytorch/botorch/pull/2908!

Balandat commented on botorch/sampling/pathwise/update_strategies.py:
```diff
@@ -137,6 +139,117 @@ def _gaussian_update_ExactGP(
+
+
+@GaussianUpdate.register(ModelListGP, LikelihoodList)
+def _gaussian_update_ModelListGP(
```
This one is completely new, right? Very nice!

Balandat commented on botorch/sampling/pathwise/utils/transforms.py:
```diff
@@ -0,0 +1,180 @@
+class SineCosineTransform(TensorTransform):
+    r"""A transform that returns concatenated sine and cosine features."""
+
+    def __init__(self, scale: Optional[Tensor] = None):
```
```suggestion
    def __init__(self, scale: Tensor | None = None):
```

Balandat commented on test/sampling/pathwise/features/test_maps.py:
```diff
@@ -6,61 +6,335 @@
+from botorch.sampling.pathwise.features import maps
+from botorch.sampling.pathwise.features.generators import gen_kernel_feature_map
+
+# Removed unused imports
+# from botorch.sampling.pathwise.utils.transforms import (
+#     ChainedTransform,
+#     FeatureSelector
+# )
```
if unused let's delete them outright - applies throughout the code
```suggestion
```

Balandat reviewed (changes requested):
I went over the main code in the PR in detail; overall this looks great, thanks for the effort and patching some gaps (e.g. `_gaussian_update_ModelListGP`). I have not reviewed the testing code in detail, but can do that after the next pass. 

The key things to address are:
1. Some additions in the patch file were not included here - curious to understand why (and if this was an oversight let's add them in - I pointed out which ones). 
2. Currently the tests still have some coverage gaps based on the codecov report [here](https://github.com/pytorch/botorch/pull/2838#issuecomment-2852070976). Please add some test cases to also cover the currently uncovered lines.

seashoo commented on botorch/sampling/pathwise/features/generators.py:
```diff
@@ -4,120 +4,144 @@
-.. [sutherland2015error]
-    D. J. Sutherland and J. Schneider. On the error of random Fourier features.
-    arXiv preprint arXiv:1506.02785 (2015).
-"""
```
My apologies, this was an oversight in a cleanup commit where I aimed to procedurally remove some cluttered developer comments. I've restored this docstring and added a few more - this is further addressed in your other comment.

seashoo commented on botorch/sampling/pathwise/features/generators.py:
```diff
@@ -4,120 +4,144 @@
-    num_inputs: int,
-    num_outputs: int,
+    num_random_features: int = 1024,
+    num_ambient_inputs: Optional[int] = None,
```
Thank you for pointing this out. I’ve implemented PEP 604 definitions throughout pathwise.

seashoo commented on botorch/sampling/pathwise/features/generators.py:
```diff
@@ -4,120 +4,144 @@
+        device=kernel.device,
+        dtype=kernel.dtype,
+    )
+    output_transforms = [transforms.SineCosineTransform(constant)]
```
Yes, I agree. The original SineCosineTransform(constant) approach was too rigid since it baked scaling into the transform class. I've implemented the conditional logic you described. Thanks for the heads up!

seashoo commented on botorch/sampling/pathwise/features/generators.py:
```diff
@@ -156,40 +181,108 @@ def _weight_generator(shape: Size) -> Tensor:
+def _gen_kernel_feature_map_multitask(
+    kernel: kernels.MultitaskKernel,
+    **kwargs: Any,
+) -> KernelFeatureMap:
```
You're right- I've updated the broader handlers to use more specific return types like `MultitaskKernelFeatureMap`, `DirectSumFeatureMap`, etc. This provides better type information to callers while maintaining runtime compatibility, since the specific types are still subclasses of `KernelFeatureMap`.

seashoo commented on botorch/sampling/pathwise/features/maps.py:
```diff
@@ -6,97 +6,383 @@
+            feature = feature_map(x, **kwargs)
+            if isinstance(feature, LinearOperator):
+                feature = feature.to_dense()
+            features.append(feature)
```
I simplified it too much in this instance. I initially went with the basic concatenation because I was having trouble getting the complex tiling/scaling logic to work correctly with all the test cases, but you're right that we need the full implementation from the patch. The simple version will break with mixed-dimension feature maps and won't handle the broadcasting/normalization correctly. I've updated the code to include the complete logic from the patch - the tiling, scaling, and advanced indexing are all necessary for mathematical correctness.

seashoo commented on botorch/sampling/pathwise/paths.py:
```diff
@@ -8,133 +8,153 @@
         self.output_transform = output_transform
-        self.paths = (
+
+        # Initialize paths dictionary - reuse ModuleDict if provided
```
Removed all attributes + redundant imports

seashoo commented:
@balandat, thank you for the comprehensive review and detailed feedback on Wilson's product_kernel.diff implementation! My apologies for the time it's taken to get back to you- some of the implementations took quite some time to fully realize and I've been balancing the work alongside my current internship. 

I'm excited to collaborate with you at a faster pace now that I've freed up time! If you have questions regarding any of my specific implementations feel free to ask in a reply to any of the comments here- I'll be able to communicate much more swiftly now. I've went ahead and resolved some of the upstream merge conflicts that appeared while I was away, and I've also filled up the code coverage gaps as you've asked.

Here's a quick summary of the major changes implemented to address your concerns:

## Mathematical Issues Resolved

### Product Kernel Implementation
Completely redesigned the `product_feature_generator` which was failing with relative errors around 0.75 vs tolerance ~0.094

- Separated finite-dimensional from infinite-dimensional sub-kernels
- Used Hadamard products for infinite-dimensional kernel combinations
- Applied outer products to merge finite-dimensional with combined infinite-dimensional kernels
- Automatically enabled `cosine_only=True` for multiple infinite-dimensional kernels to avoid problematic tensor products

### Transform System Overhaul
Fixed scaling and coordination issues across the transform pipeline

- **SineCosineTransform**: Added conditional logic instead of rigid constant scaling 
- **Parameter Passing**: Improved explicit `num_ambient_inputs` handling vs complex kwargs manipulation
- **Transform Chaining**: Enhanced `append_transform` utility to properly handle None cases 

## Architectural Improvements

### Feature Map Redesign
Built comprehensive feature map architecture

- **DirectSumFeatureMap**: Rewrote `raw_output_shape` method to handle mixed dimensionality without MagicMock detection 
- **SparseDirectSumFeatureMap**: Implemented for completeness, available for manual use 
- **HadamardProductFeatureMap/OuterProductFeatureMap**: Enhanced for proper kernel composition

### Code Organization
Restructured from monolithic `utils.py` into modular package

- Split into dedicated helpers, mixins, and transforms modules
- Improved dispatcher system with specific return types (`MultitaskKernelFeatureMap`, `DirectSumFeatureMap`, etc.)
- Fixed `sub_kernels` parameter usage for LCM kernel compatibility

## Code Quality Enhancements

### Modern Python Standards

- Adopted PEP 604 union syntax (`A | B` vs `Union[A, B]`) throughout pathwise directory 
- Replaced large comment blocks with Google-style docstrings 
- Cleaned up redundant imports and unused attributes 

### Type Safety
Enhanced type annotations and return type specificity for better IDE support

Technical details regarding the issues + approaches taken in response to your suggestions are further addressed in my replies!

Balandat commented:
Thanks for all the updates, @seashoo. Sorry for the delay on the review; I've been traveling for a conference and taking some time off. I should be able to provide a detailed review next week.

hvarfner commented on botorch/sampling/pathwise/prior_samplers.py:
```diff
@@ -102,36 +117,108 @@ def _draw_kernel_feature_paths_ExactGP(
+    # containing data_covar_module * task_covar_module
+    from gpytorch.kernels import ProductKernel
+
+    if isinstance(model.covar_module, ProductKernel):
```
The MTGPs (including the fully Bayesian variant) now use a ProductKernel by definition, so this check and the conditional logic should be obsolete.

hvarfner replied:
And if there are some edge cases where we don't adhere to the standard `ProductKernel(IndexKernel, SomeOtherKernel)` where both have the expected `active_dims`, I think it's best to raise an error!

hvarfner commented on botorch/sampling/pathwise/update_strategies.py:
```diff
@@ -143,6 +144,160 @@ def _gaussian_update_ExactGP(
+    # containing data_covar_module * task_covar_module
+    from gpytorch.kernels import ProductKernel
+
+    if isinstance(model.covar_module, ProductKernel):
```
Once again, this conditional logic should not be needed anymore!

hvarfner commented on botorch/sampling/pathwise/utils/helpers.py:
```diff
@@ -0,0 +1,333 @@
+INF_DIM_KERNELS: Tuple[Type[Kernel], ...] = (
+    kernels.MaternKernel,
+    kernels.RBFKernel,
+    kernels.MultitaskKernel,
```
Why is `MultiTaskKernel` necessarily in here? Shouldn't `IndexKernel` be here too in that case?

seashoo replied:
`INF_DIM_KERNELS` is the “shortcut” list for kernels that are always treated as infinite-dimensional when deciding whether we need random Fourier features etc.  `RBFKernel` and `MaternKernel` are on the list because their exact feature map is infinite.  `IndexKernel` itself is finite-dimensional (exactly `num_tasks` features), so we deliberately keep it out of the list.

seashoo commented on botorch/sampling/pathwise/features/maps.py:
```diff
@@ -6,31 +6,383 @@
+        self.input_transform = input_transform
+        self.output_transform = output_transform
+
+    def forward(self, x: Tensor, **kwargs: Any) -> Tensor:
```
Thanks for the feedback- I’ve added in-line comments to DirectSumFeatureMap.forward (from the tiling logic through the final concatenation) to clarify:

- Why I collect/scale individual feature blocks.
- How the tiling & broadcasting works for lower-dimensional feature maps.
- The rationale behind the rescaling.
- How the multi_index trick avoids extra allocations.

Let me know if any sections still feel unclear or if you’d like similar commentary elsewhere.

seashoo commented on botorch/sampling/pathwise/features/maps.py:
```diff
@@ -6,31 +6,383 @@
+        return torch.concat(blocks, dim=-1)
+
+    @property
+    def raw_output_shape(self) -> Size:
```
`raw_output_shape` does two things in one go:  

1. For every sub-map it first figures out the shape you would get after tiling / broadcasting inside `forward` (e.g. when a 1-D feature map has to be expanded to align with a 2-D one).  
2. It then concatenates those shapes along the feature dimension, so the final size of the last axis is the sum of the sub-maps’ feature counts.  

`torch.broadcast_shapes` by itself only answers “what common shape can all these tensors be viewed as without copying?”.  It never alters the size of any dimension- especially not the last one.  

In our case we need to grow the last dimension because we’re gluing feature vectors together; that’s why `raw_output_shape` can’t be expressed as a single `torch.broadcast_shapes` call.

TaiwoAdebiyi23 commented on botorch/sampling/pathwise/features/generators.py:
```diff
@@ -53,71 +78,109 @@ def gen_kernel_features(
+    num_random_features: int,
+    num_inputs: int | None = None,
+    random_feature_scale: float | None = None,
+    cosine_only: bool = False,
```
Hi @hvarfner, I believe @seashoo followed the direct implementation from the original patch provided by Balandat, likely implemented by Wilson to make computation tractable for product of infinite dimensional kernels. By enforcing `cosine_only == True`, we ensure that the feature map takes the element-wise product of the features. Otherwise, we would need to take the tensor product of each pair of sine and cosine features.
 
I don't think we have a specific reference that comments on this edge case, but as you already know, `cosine_only` specifies whether to use cosine features with a random phase, instead of paired sine and cosine features, as described in Rahimi & Recht (2007) on Random Fourier Features (RFF) and Sutherland (2015) on RFF error bounds. I'll let you know if I come across any further references that clarify this implementation.

hvarfner commented on botorch/sampling/pathwise/features/maps.py:
```diff
@@ -6,31 +6,383 @@
+
+    @property
+    def raw_output_shape(self) -> Size:
+        if not self:
```
Terrific, thanks!

hvarfner commented on botorch/sampling/pathwise/update_strategies.py:
```diff
@@ -210,16 +220,15 @@ def _draw_kernel_feature_paths_MultiTaskGP(
+        if data_kernel is None:
+            raise ValueError(
+                f"Could not identify data kernel from ProductKernel. "
+                "MTGPs should follow the standard ProductKernel(IndexKernel, SomeOtherKernel) pattern."
```
And here

hvarfner reviewed (commented):
I reviewed the changes, and these look good in isolation - just some minor nits. I'll go through thing again, given that it was a while ago that we last looked at it.

hvarfner commented on botorch/sampling/pathwise/features/maps.py:
```diff
@@ -105,29 +481,157 @@ def __init__(
-    def batch_shape(self) -> Size:
-        return self.kernel.batch_shape
+    def raw_output_shape(self) -> Size:
+        return self.kernel.raw_var.shape[-1:]
```
I could see the raw_var not always being on the `IndexKernel`, or have it be optional - we're looking into that ATM. Can you make it depend on the `covar_matrix` instead?

hvarfner reviewed (commented):
This looks great, extraordinary work. Seems like the branch is not entirely in sync, given the changes in `optim`. Thanks a lot.

I think the `get_train_inputs` and `get_train_targets` methods and dispatching logic is helpful in general, and should probably be located models/utils/helpers.py. If it's not too much work, it would be nice to have it there instead.

seashoo commented:
> Thanks!
> 
> > merging with BoTorch's upstream is causing dependency issues with GPyTorch imports on my local. I will investigate this issue further and commit the working upstream in the coming days.
> 
> @seashoo when you say dependency issues, do you mean in terms of installing the package or in terms of causing (e.g. import) failures? There has been a new gpytorch release just recently that you may need to update to: https://github.com/cornellius-gp/gpytorch/releases/tag/v1.14.2

My local + virtual environment weren't taking the new update at the time. I ran a fresh install of gpytorch + all the necessary development tools in a new python venv- should be good now!
